tapdb/assets_store_test.go

```diff
@@ -2349,3 +2349,109 @@ func TestQueryAssetBalances(t *testing.T) {
 	}
 	require.Equal(t, totalGroupedBalances, leasedBalanceByGroupSum)
 }
+
+func TestQueryAssetBalancesCustomChannelFunding(t *testing.T) {
+	t.Parallel()
+
+	_, assetsStore, _ := newAssetStore(t)
+	ctx := context.Background()
+
+	// First, we'll generate 2 assets, one of them having a script key that
+	// is the typical funding script key.
+	const numAssets = 2
+	const numGroups = 1
+	assetGen := newAssetGenerator(t, numAssets, numGroups)
+
+	fundingScriptTree := newFundingScriptTree()
+	fundingScriptKey := asset.ScriptKey{
+		PubKey: fundingScriptTree.TaprootKey,
+	}
+
+	assetDesc := []assetDesc{
+		{
+			assetGen:    assetGen.assetGens[0],
+			anchorPoint: assetGen.anchorPoints[0],
+			keyGroup:    assetGen.groupKeys[0],
+			amt:         4,
+			scriptKey:   &fundingScriptKey,
+		},
+		{
+			assetGen:    assetGen.assetGens[1],
+			anchorPoint: assetGen.anchorPoints[1],
+			keyGroup:    assetGen.groupKeys[0],
+			amt:         4,
+		},
+	}
+	assetGen.genAssets(t, assetsStore, assetDesc)
+
+	// Hit both balance queries, they should return the same result.
+	includeLeased := false
+	balances, err := assetsStore.QueryBalancesByAsset(
+		ctx, nil, includeLeased,
+	)
+	require.NoError(t, err)
+	balancesByGroup, err := assetsStore.QueryAssetBalancesByGroup(
+		ctx, nil, includeLeased,
+	)
+	require.NoError(t, err)
+	require.Len(t, balances, numAssets-1)
+	require.Len(t, balancesByGroup, numAssets-1)
+
+	// Both sums should be equal to the amount of the second asset; the
+	// asset that is not part of a custom channel funding tx.
+	balanceSum := uint64(0)
+	for _, balance := range balances {
+		balanceSum += balance.Balance
+	}
+	require.Equal(t, assetDesc[1].amt, balanceSum)
+
+	balanceByGroupSum := uint64(0)
+	for _, balance := range balancesByGroup {
+		balanceByGroupSum += balance.Balance
+	}
+	require.Equal(t, assetDesc[1].amt, balanceByGroupSum)
+}
+
+// newFundingScriptTree creates a new funding script tree for a custom channel
+// asset-level script key. The script tree is constructed with a simple OP_TRUE
+// script that allows anyone to spend the output. This simplifies the funding
+// process as no signatures for the asset-level witnesses need to be exchanged.
+// This is still safe because the BTC level multi-sig output is still protected
+// by a 2-of-2 MuSig2 output.
+func newFundingScriptTree() *FundingScriptTree {
```

Comment:
Instead of creating a duplicate here, let's just move the original type to `tapscript`, which _should_ allow us to access it from `tapdb` without creating a circular package dependency.